server/src/main/java/org/elasticsearch/indices/ShardLimitValidator.java

```diff
@@ -81,55 +111,119 @@ public void validateShardLimit(final Settings settings, final ClusterState state
     }
 
     // package-private for testing
-    static Optional<String> checkShardLimit(int newShards, ClusterState state, int maxShardsPerNodeSetting) {
-        int nodeCount = state.getNodes().getDataNodes().size();
-
+    static Optional<String> checkShardLimit(int newShards, ClusterState state, int maxShardsPerNode, int nodeCount, String group) {
         // Only enforce the shard limit if we have at least one data node, so that we don't block
         // index creation during cluster setup
-        if (nodeCount == 0 || newShards < 0) {
+        if (nodeCount == 0 || newShards <= 0) {
             return Optional.empty();
         }
-        int maxShardsPerNode = maxShardsPerNodeSetting;
         int maxShardsInCluster = maxShardsPerNode * nodeCount;
         int currentOpenShards = state.getMetadata().getTotalOpenIndexShards();
 
         if ((currentOpenShards + newShards) > maxShardsInCluster) {
-            String errorMessage = "this action would add [" + newShards + "] total shards, but this cluster currently has [" +
-                currentOpenShards + "]/[" + maxShardsInCluster + "] maximum shards open";
-            return Optional.of(errorMessage);
+            Predicate<IndexMetadata> indexMetadataPredicate = imd ->
+                imd.getState().equals(IndexMetadata.State.OPEN) && group.equals(INDEX_SETTING_SHARD_LIMIT_GROUP.get(imd.getSettings()));
+            long currentFilteredShards = StreamSupport.stream(state.metadata().indices().values().spliterator(), false).map(oc -> oc.value)
+                .filter(indexMetadataPredicate).mapToInt(IndexMetadata::getTotalNumberOfShards).sum();
+            if ((currentFilteredShards + newShards) > maxShardsInCluster) {
+                String errorMessage = "this action would add [" + newShards + "] shards, but this cluster currently has [" +
+                    currentFilteredShards + "]/[" + maxShardsInCluster + "] maximum " + group + " shards open";
+                return Optional.of(errorMessage);
+            }
         }
         return Optional.empty();
     }
+
+    private static int nodeCount(ClusterState state, Predicate<DiscoveryNode> nodePredicate) {
+        return (int)
+            StreamSupport.stream(state.getNodes().getDataNodes().values().spliterator(), false)
+                .map(oc -> oc.value).filter(nodePredicate).count();
+    }
+
+    private static boolean hasFrozen(DiscoveryNode node) {
+        final List<DiscoveryNodeRole> dataRoles =
+            node.getRoles().stream().filter(DiscoveryNodeRole::canContainData).collect(Collectors.toUnmodifiableList());
+
+        return dataRoles.contains(DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE);
+    }
+
+    private static boolean hasNonFrozen(DiscoveryNode node) {
+        final List<DiscoveryNodeRole> dataRoles =
+            node.getRoles().stream().filter(DiscoveryNodeRole::canContainData).collect(Collectors.toUnmodifiableList());
```

Comment:
`return node.getRoles().stream().anyMatch(r -> r.canContainData() && r !=DiscoveryNodeRole.DATA_FROZEN_NODE_ROLE);`?

Reply:
This suggested predicate almost exactly matches the name of the method (it has a non-frozen data role), while the current implementation requires a little bit of though. And we can avoid the unnecessary materialization of a collection, etc.

Reply:
Thanks, that is nicer, added in 67e92fdfad17a875639dcdc3b2dd41f98b9bf8fe